### PR TITLE
[DataPipe] Traverse DataPipe graph excluding primitive and callable

### PR DESCRIPTION
**Pull request by ejguan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#67783 [DataPipe] Traverse DataPipe graph excluding primitive and callable**

Add `getstate_hook` to exclude primitive objects and callable when serialization when `exclude_primitive` is enabled for `traverse`.
For graph traversing, we don't have to handle the lambda and other stuff.
- This is going to solve a problem that generator objects would also break the `traverse`
  -  Some `IterDataPipe` instances like `Demux` holds a generator object (normally iterator) when It starts iteration. https://github.com/pytorch/pytorch/blob/89b02fc70b6a279b602545fddbcd56b4248cee4f/torch/utils/data/datapipes/iter/combining.py#L220
  - If users start to iterate datapipe first for validation purpose, the generator object is created during iteration. Then, when they send the pipeline into `DataLoader`, the `traverse` is going to break.
- This is also used by `OnDiskCacheHolder` to trace the DataPipe Graph without worry about lambda and generator.

Differential Revision: [D32146697](https://our.internmc.facebook.com/intern/diff/D32146697)

cc @VitalyFedyunin @ejguan @NivekT

### CODE REVIEW

File: torch/utils/data/dataset.py [hunk 1]
```diff
@@ -225,6 +226,11 @@ def __getattr__(self, attribute_name):
         else:
             raise AttributeError
 
+    def __getstate__(self):
+        if IterableDataset.getstate_hook is not None:
+            return IterableDataset.getstate_hook(self)
+        return self.__dict__
+
```
Comment (ejguan) [inline torch/utils/data/dataset.py:233]: I really want these APIs moved out of `IterableDataset` by making `IterDataPipe` derived from `IterableDataset`

File: torch/utils/data/datapipes/iter/callable.py [hunk 2]
```diff
@@ -151,6 +151,9 @@ def __len__(self) -> int:
         )
 
     def __getstate__(self):
+        if IterDataPipe.getstate_hook is not None:
+            return IterDataPipe.getstate_hook(self)
+
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/callable.py:156]: This is needed for all subclasses of `IterDataPipe` which implement `__getstate__`

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32146697).

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32146697).

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32146697).

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32146697).

**Review by VitalyFedyunin — APPROVED** [review]
Looks good, but `set_getstate_hook(None)` little bit counterintuitive, maybe do `disable_getstate_hook`

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32146697).

**Pull request closed without merging**